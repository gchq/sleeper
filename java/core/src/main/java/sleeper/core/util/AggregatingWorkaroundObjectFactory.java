/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import sleeper.core.iterator.AggregationFilteringIterator;
import sleeper.core.iterator.SortedRecordIterator;
import sleeper.core.properties.model.CompactionMethod;

public class AggregatingWorkaroundObjectFactory {
    private final ObjectFactory inner;

    public AggregatingWorkaroundObjectFactory(ObjectFactory inner) {
        this.inner = inner;
    }

    public <T> T getObject(String className, Class<T> parentClass) throws ObjectFactoryException {
        if (parentClass.equals(SortedRecordIterator.class) && className.equalsIgnoreCase(CompactionMethod.AGGREGATION_ITERATOR_NAME)) {
            return inner.getObject(AggregationFilteringIterator.class.getName(), parentClass);
        } else {
            return inner.getObject(className, parentClass);
        }
    }

}
