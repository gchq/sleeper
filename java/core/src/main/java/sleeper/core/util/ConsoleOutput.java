/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import java.io.PrintStream;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;

/**
 * This class is used to print out to the console.
 */
public class ConsoleOutput {

    public static final String CLEAR_CONSOLE = "\033[H\033[2J";

    private final PrintStream out;

    public ConsoleOutput(PrintStream out) {
        this.out = out;
    }

    /**
     * This method clears the console screen then prints the input on a new line.
     *
     * @param message String message to be printed after screen is cleared.
     */
    public void clearScreen(String message) {
        out.print(CLEAR_CONSOLE);
        out.flush();
        out.println(message);
    }

    /**
     * This method prints the input message on a new line.
     *
     * @param message String message to be printed on a new line.
     */
    public void println(String message) {
        out.println(message);
    }

    /**
     * This method prints a blank line.
     */
    public void println() {
        out.println();
    }

    /**
     * This method prints to the console using a formatter.
     *
     * @param format String format to be used for the message.
     * @param args   A number of args to be passed into the formatter.
     */
    public void printf(String format, Object... args) {
        out.printf(format, args);
    }

    /**
     * This method returns a new PrintWriter with the PrintStream of this object, autoFlush set to true and
     * StandardCharsets.UTF_8.
     *
     * @return A new PrintWriter with the PrintStream of this object, autoFlush set to true and
     *         StandardCharsets.UTF_8.
     */
    public PrintWriter writer() {
        return new PrintWriter(out, true, StandardCharsets.UTF_8);
    }

    /**
     * This method returns the PrintStream object assigned to the out variable.
     *
     * @return The PrintStream object assigned to the out variable.
     */
    public PrintStream printStream() {
        return out;
    }
}
