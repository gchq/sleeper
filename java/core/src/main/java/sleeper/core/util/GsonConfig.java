/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.core.util;

import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;
import java.time.Instant;

/**
 * A helper class for common GSON configuration for JSON internal to Sleeper.
 */
public class GsonConfig {

    private GsonConfig() {
    }

    /**
     * Creates a GSON builder preconfigured for JSON internal to Sleeper. This must be further configured to work with
     * specific Sleeper entities.
     *
     * @return the new builder
     */
    public static GsonBuilder standardBuilder() {
        return new GsonBuilder().serializeSpecialFloatingPointValues()
                .registerTypeAdapter(Instant.class, new InstantSerDe());
    }

    /**
     * A GSON plugin to serialise/deserialise an instant.
     */
    private static class InstantSerDe implements JsonSerializer<Instant>, JsonDeserializer<Instant> {
        @Override
        public Instant deserialize(JsonElement element, Type type, JsonDeserializationContext context) throws JsonParseException {
            return Instant.ofEpochMilli(element.getAsLong());
        }

        @Override
        public JsonElement serialize(Instant instant, Type type, JsonSerializationContext context) {
            return new JsonPrimitive(instant.toEpochMilli());
        }
    }
}
