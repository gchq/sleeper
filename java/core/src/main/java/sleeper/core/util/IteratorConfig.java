/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

/** Config class for getting iterator's from the iterator factory. */
public class IteratorConfig {
    private final String iteratorClassName;
    private final String iteratorConfigString;
    private final String filters;
    private final String aggregationString;

    public IteratorConfig(Builder builder) {
        this.iteratorClassName = builder.iteratorClassName;
        this.iteratorConfigString = builder.iteratorConfigString;
        this.filters = builder.filters;
        this.aggregationString = builder.aggregationString;
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getIteratorClassName() {
        return iteratorClassName;
    }

    public String getIteratorConfigString() {
        return iteratorConfigString;
    }

    public String getFilters() {
        return filters;
    }

    public String getAggregationString() {
        return aggregationString;
    }

    /**
     * Checks if a iterator should be applied based on if a class name or filters have been set.
     *
     * @return true if the iterator should be applied, otherwise false
     */
    public boolean shouldIteratorBeApplied() {
        return iteratorClassName != null || filters != null || aggregationString != null;
    }

    /**
     * Builder for iterator config object.
     */
    public static final class Builder {
        private String iteratorClassName;
        private String iteratorConfigString;
        private String filters;
        private String aggregationString;

        private Builder() {
        }

        /**
         * Sets the iterator class name.
         *
         * @param  iteratorClassName the name of the iterator class to build
         * @return                   builder for method chaining
         */
        public Builder iteratorClassName(String iteratorClassName) {
            this.iteratorClassName = iteratorClassName;
            return this;
        }

        /**
         * Sets the iterator config string.
         *
         * @param  iteratorConfigString the config string to be used for the iterator
         * @return                      builder for method chaining
         */
        public Builder iteratorConfigString(String iteratorConfigString) {
            this.iteratorConfigString = iteratorConfigString;
            return this;
        }

        /**
         * Sets the filters string.
         *
         * @param  filters the filters string to be used for the iterator
         * @return         builder for method chaining
         */
        public Builder filters(String filters) {
            this.filters = filters;
            return this;
        }

        /**
         * Sets the aggregation string.
         *
         * @param  aggregationString the aggregation string to be used for the iterator
         * @return                   builder for method chaining
         */
        public Builder aggregationString(String aggregationString) {
            this.aggregationString = aggregationString;
            return this;
        }

        public IteratorConfig build() {
            return new IteratorConfig(this);
        }
    }
}
