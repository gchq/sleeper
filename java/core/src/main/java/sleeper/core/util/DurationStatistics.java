/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import java.time.Duration;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toUnmodifiableList;

/**
 * Calculates statistics about a dataset of Duration objects.
 */
public class DurationStatistics {
    private final Duration min;
    private final Duration mean;
    private final Duration max;
    private final Duration standardDeviation;

    public DurationStatistics(Builder builder) {
        this.min = builder.min;
        this.mean = builder.mean;
        this.max = builder.max;
        this.standardDeviation = builder.standardDeviation;
    }

    /**
     * Calculates statistics from a stream of Duration objects, if any are present.
     *
     * @param  durations the durations
     * @return           the statistics, if any are present
     */
    public static Optional<DurationStatistics> fromIfAny(Stream<Duration> durations) {
        List<Long> durationsInMills = durations.map(Duration::toMillis).collect(toUnmodifiableList());
        if (durationsInMills.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(builder()
                    .computeFromMilliseconds(durationsInMills)
                    .build());
        }
    }

    private static Builder builder() {
        return new Builder();
    }

    @Override
    public String toString() {
        return String.format("avg: %s, min: %s, max: %s, std dev: %s",
                LoggedDuration.withShortOutput(mean),
                LoggedDuration.withShortOutput(min),
                LoggedDuration.withShortOutput(max),
                LoggedDuration.withShortOutput(standardDeviation));
    }

    /**
     * Builds instances of this class.
     */
    private static class Builder {
        private Duration min;
        private Duration mean;
        private Duration max;
        private Duration standardDeviation;
        private long minMillis = Long.MAX_VALUE;
        private long maxMillis = Long.MIN_VALUE;

        Builder computeFromMilliseconds(List<Long> durationsInMilliseconds) {
            int n = durationsInMilliseconds.size();
            double meanMillis = durationsInMilliseconds.stream()
                    .peek(millis -> minMillis = Math.min(millis, minMillis))
                    .peek(millis -> maxMillis = Math.max(millis, maxMillis))
                    .mapToLong(millis -> millis).sum() / (double) n;
            double variance = durationsInMilliseconds.stream()
                    .mapToDouble(millis -> Math.pow(millis - meanMillis, 2))
                    .sum() / n;
            min = Duration.ofMillis(minMillis);
            mean = Duration.ofMillis((long) meanMillis);
            max = Duration.ofMillis(maxMillis);
            standardDeviation = Duration.ofMillis((long) Math.sqrt(variance));
            return this;
        }

        DurationStatistics build() {
            return new DurationStatistics(this);
        }
    }
}
