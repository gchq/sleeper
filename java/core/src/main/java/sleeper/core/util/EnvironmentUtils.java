/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import sleeper.core.properties.instance.InstanceProperties;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.CONFIG_BUCKET;
import static sleeper.core.properties.instance.LoggingLevelsProperty.APACHE_LOGGING_LEVEL;
import static sleeper.core.properties.instance.LoggingLevelsProperty.AWS_LOGGING_LEVEL;
import static sleeper.core.properties.instance.LoggingLevelsProperty.LOGGING_LEVEL;
import static sleeper.core.properties.instance.LoggingLevelsProperty.PARQUET_LOGGING_LEVEL;
import static sleeper.core.properties.instance.LoggingLevelsProperty.ROOT_LOGGING_LEVEL;

/**
 * Utilities to set environment variables during deployment.
 */
public class EnvironmentUtils {

    private EnvironmentUtils() {
        // Prevents instantiation
    }

    /**
     * Creates the default environment for a deployed process running in a Sleeper instance.
     *
     * @param  instanceProperties the instance properties
     * @return                    the environment variables
     */
    public static Map<String, String> createDefaultEnvironment(InstanceProperties instanceProperties) {
        Map<String, String> environmentVariables = createDefaultEnvironmentNoConfigBucket(instanceProperties);
        environmentVariables.put(CONFIG_BUCKET.toEnvironmentVariable(),
                instanceProperties.get(CONFIG_BUCKET));
        return environmentVariables;
    }

    /**
     * Creates the default environment for a deployed process running in a Sleeper instance, without the config bucket
     * environment variable.
     *
     * @param  instanceProperties the instance properties
     * @return                    the environment variables
     */
    public static Map<String, String> createDefaultEnvironmentNoConfigBucket(InstanceProperties instanceProperties) {
        Map<String, String> environmentVariables = new HashMap<>();
        environmentVariables.put("JAVA_TOOL_OPTIONS", createToolOptions(instanceProperties));
        return environmentVariables;
    }

    private static String createToolOptions(InstanceProperties instanceProperties) {
        StringBuilder sb = new StringBuilder();
        Stream.of(LOGGING_LEVEL,
                ROOT_LOGGING_LEVEL,
                APACHE_LOGGING_LEVEL,
                PARQUET_LOGGING_LEVEL,
                AWS_LOGGING_LEVEL)
                .filter(instanceProperties::isSet)
                .forEach(s -> sb.append("-D").append(s.getPropertyName())
                        .append("=").append(instanceProperties.get(s)).append(" "));
        Stream.of("java.base/java.nio=ALL-UNNAMED",
                "java.base/sun.nio.ch=ALL-UNNAMED",
                "java.base/java.util=ALL-UNNAMED",
                "java.base/java.lang.invoke=ALL-UNNAMED")
                .forEach(s -> sb.append("--add-opens=").append(s).append(" "));
        return sb.toString();
    }
}
