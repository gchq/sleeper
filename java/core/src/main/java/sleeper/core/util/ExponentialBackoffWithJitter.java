/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.Objects;
import java.util.function.DoubleSupplier;

/**
 * Waits for increasing periods of time during retries. Uses exponential backoff with full jitter. See the below
 * article.
 * <p>
 * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
 */
public class ExponentialBackoffWithJitter {
    public static final Logger LOGGER = LoggerFactory.getLogger(ExponentialBackoffWithJitter.class);

    private final DoubleSupplier randomJitterFraction;
    private final Waiter waiter;
    private final WaitRange waitRange;

    public ExponentialBackoffWithJitter(WaitRange waitRange, DoubleSupplier randomJitterFraction, Waiter waiter) {
        this.waitRange = Objects.requireNonNull(waitRange, "waitRange must not be null");
        this.randomJitterFraction = Objects.requireNonNull(randomJitterFraction, "randomJitterFraction must not be null");
        this.waiter = Objects.requireNonNull(waiter, "waiter must not be null");
    }

    public ExponentialBackoffWithJitter(WaitRange waitRange) {
        this(waitRange, Math::random, Thread::sleep);
    }

    /**
     * Waits for a time calculated from the number of attempts.
     *
     * @param  attempt              the number of the attempt that is about to be made, starting at 1
     * @return                      the number of milliseconds waited for
     * @throws InterruptedException if the current thread was interrupted
     */
    public long waitBeforeAttempt(int attempt) throws InterruptedException {
        if (attempt == 1) { // No wait on first attempt
            return 0;
        } else if (attempt < 1) {
            throw new IllegalArgumentException("Attempt count must start at 1, found " + attempt);
        }
        long waitMillis = getWaitMillisBeforeAttempt(attempt);
        LOGGER.debug("Sleeping for {}", LoggedDuration.withFullOutput(Duration.ofMillis(waitMillis)));
        waiter.waitForMillis(waitMillis);
        return waitMillis;
    }

    private long getWaitMillisBeforeAttempt(int attempt) {
        double waitCeilingInSeconds = Math.min(
                waitRange.maxWaitCeilingSecs,
                waitRange.firstWaitCeilingSecs * Math.pow(2.0, attempt - 2)); // First retry is attempt 2
        return (long) (randomJitterFraction.getAsDouble() * waitCeilingInSeconds * 1000L);
    }

    /**
     * Defines a range for the wait time ceiling. The ceiling is an amount of time that increases exponentially for
     * each retry. Jitter is then applied to this ceiling to produce the actual wait time.
     * <p>
     * The ceiling starts at a first value for the first retry, then increases exponentially until it reaches a maximum.
     */
    public static class WaitRange {
        private final double firstWaitCeilingSecs;
        private final double maxWaitCeilingSecs;

        private WaitRange(double firstWaitCeilingSecs, double maxWaitCeilingSecs) {
            this.firstWaitCeilingSecs = firstWaitCeilingSecs;
            this.maxWaitCeilingSecs = maxWaitCeilingSecs;
        }

        /**
         * Instantiates this class.
         *
         * @param  firstWaitCeilingSecs the wait ceiling in seconds for the first retry, before applying jitter
         * @param  maxWaitCeilingSecs   the maximum wait ceiling in seconds, before applying jitter
         * @return                      an instance of this class
         */
        public static WaitRange firstAndMaxWaitCeilingSecs(double firstWaitCeilingSecs, double maxWaitCeilingSecs) {
            return new WaitRange(firstWaitCeilingSecs, maxWaitCeilingSecs);
        }

        public Duration getFirstWaitCeiling() {
            return Duration.ofMillis((long) (firstWaitCeilingSecs * 1000));
        }

        public Duration getMaxWaitCeiling() {
            return Duration.ofMillis((long) (maxWaitCeilingSecs * 1000));
        }
    }
}
