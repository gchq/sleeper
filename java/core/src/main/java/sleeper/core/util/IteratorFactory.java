/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.util;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.core.iterator.AggregationFilteringIterator;
import sleeper.core.iterator.ConfigStringAggregationFilteringIterator;
import sleeper.core.iterator.ConfigStringIterator;
import sleeper.core.iterator.FilterAggregationConfig;
import sleeper.core.iterator.IteratorCreationException;
import sleeper.core.iterator.SortedRowIterator;
import sleeper.core.properties.model.DataEngine;
import sleeper.core.schema.Schema;

import java.util.List;
import java.util.Locale;
import java.util.Optional;

/**
 * A factory class for Sleeper iterators.
 */
public class IteratorFactory {
    private static final Logger LOGGER = LoggerFactory.getLogger(IteratorFactory.class);

    private final ObjectFactory inner;

    public IteratorFactory(ObjectFactory inner) {
        this.inner = inner;
    }

    /**
     * Creates an initialises an iterator.
     *
     * The named iterator class is created and its {@link ConfigStringIterator#init(String, Schema)} method is called.
     * If the special keyword for aggregation is specified as the class name
     * {@link DataEngine#AGGREGATION_ITERATOR_NAME},
     * then an aggregation iterator is created and initialised.
     *
     * @param  iteratorConfig            config object holding iterator details
     * @return                           an initialised iterator
     * @throws IteratorCreationException if an iterator can't be created, for example it's class definition can't be
     *                                   found
     * @see                              ConfigStringAggregationFilteringIterator
     */
    public SortedRowIterator getIterator(IteratorConfig iteratorConfig) throws IteratorCreationException {
        try {
            if (iteratorConfig.getFilters() == null) {
                ConfigStringIterator iterator;
                String className = iteratorConfig.getIteratorClassName();

                // If aggregation keyword is used, create specific iterator
                if (className.equalsIgnoreCase(DataEngine.AGGREGATION_ITERATOR_NAME)) {
                    iterator = new ConfigStringAggregationFilteringIterator();
                } else {
                    iterator = inner.getObject(className, ConfigStringIterator.class);
                }
                LOGGER.debug("Created iterator of class {}", className);
                iterator.init(iteratorConfig.getIteratorConfigString(), iteratorConfig.getSchema());
                LOGGER.debug("Initialised iterator with config {}", iteratorConfig.getIteratorConfigString());
                return iterator;
            } else {
                AggregationFilteringIterator iterator = new AggregationFilteringIterator();
                iterator.setFilterAggregationConfig(getConfigFromProperties(iteratorConfig));
                iterator.setSchema(iteratorConfig.getSchema());
                return iterator;
            }
        } catch (ObjectFactoryException exc) {
            throw new IteratorCreationException(exc);
        }
    }

    /**
     * Gets the filter aggregation config when set by table properties.
     * Currently just gets the filter ageoff(column,age).
     * Currently only uses filters from input but eventually will build aggregations from it too.
     *
     * @param  iteratorConfig config for an iterator that should have filters set in it
     * @return                filter aggregation config to be used in an iterator
     */
    private FilterAggregationConfig getConfigFromProperties(IteratorConfig iteratorConfig) {
        String[] filterParts = iteratorConfig.getFilters().split("\\(");
        if ("ageoff".equals(filterParts[0].toLowerCase(Locale.ENGLISH))) {
            String[] filterInput = StringUtils.chop(filterParts[1]).split(","); //Remove the trailing ')'
            Optional<String> filterColumn = Optional.of(filterInput[0]);
            long maxAge = Long.parseLong(filterInput[1]);
            return new FilterAggregationConfig(List.of(), filterColumn, maxAge, List.of());
        } else {
            throw new IllegalStateException("Sleeper table filter not set to match ageOff(column,age), was: " + filterParts[0]);
        }
    }
}
