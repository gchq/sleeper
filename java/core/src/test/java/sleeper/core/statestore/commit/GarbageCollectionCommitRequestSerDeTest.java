/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.commit;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class GarbageCollectionCommitRequestSerDeTest {

    private GarbageCollectionCommitRequestSerDe serDe = new GarbageCollectionCommitRequestSerDe();

    @Test
    void shouldSerialiseGarbageCollectionCommitRequest() {
        // Given
        GarbageCollectionCommitRequest garbageCollectionCommitRequest = new GarbageCollectionCommitRequest("tableId",
                List.of("file1.parquet", "file2.parquet", "file3.parquet"));

        // When
        String json = serDe.toJsonPrettyPrint(garbageCollectionCommitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(garbageCollectionCommitRequest);
        Approvals.verify(json);
    }

    @Test
    void shouldFailToDeserialiseNonGarbageCollectionCommitRequest() {
        assertThatThrownBy(() -> serDe.fromJson("{\"type\": \"OTHER\", \"request\":{}}"))
                .isInstanceOf(IllegalArgumentException.class);
    }
}
