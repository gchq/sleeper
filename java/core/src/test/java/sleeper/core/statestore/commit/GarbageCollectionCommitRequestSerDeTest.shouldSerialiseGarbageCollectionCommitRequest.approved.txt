{
  "type": "GARBAGE_COLLECTED_FILES",
  "request": {
    "tableId": "tableId",
    "filenames": [
      "file1.parquet",
      "file2.parquet",
      "file3.parquet"
    ]
  }
}